Src/ILGPU/Runtime/ExchangeBuffer.cs

```diff
@@ -64,6 +65,82 @@ public static class ExchangeBuffer
+        /// same size in pinned CPU memory to enable asynchronous memory transfers
+        /// </summary>
+        /// <typeparam name="T">The element type.</typeparam>
+        /// <param name="accelerator">The associated accelerator to use.</param>
+        /// <param name="extent">The extent (number of elements to allocate).</param>
+        /// <param name="mode">The exchange buffer mode to use.</param>
+        /// <returns>The allocated 2D exchange buffer.</returns>
+        public static ExchangeBuffer2D<T> AllocateExchangeBuffer<T> (
+            this Accelerator accelerator,
+            Index2 extent,
+            ExchangeBufferMode mode)
+            where T : unmanaged
+        {
+            var gpuBuffer = accelerator.Allocate<T>(extent);
+            return new ExchangeBuffer2D<T>(gpuBuffer, mode);
+        }
+
+        /// <summary>
+        /// Allocates a new 3D exchange buffer that allocates the specified amount of
+        /// elements on the current accelerator. Furthermore, it keeps a buffer of the
+        /// same size in pinned CPU memory to enable asynchronous memory transfers
+        /// </summary>
+        /// <typeparam name="T">The element type.</typeparam>
+        /// <param name="accelerator">The associated accelerator to use.</param>
+        /// <param name="extent">The extent (number of elements to allocate).</param>
+        /// <returns>The allocated 3D exchange buffer.</returns>
+        /// <remarks>
+        /// This function uses the default buffer allocation mode
+        /// <see cref="ExchangeBufferMode.PreferPagedLockedMemory"/>
+        /// </remarks>
+        public static ExchangeBuffer3D<T> AllocateExchangeBuffer<T>(
+            this Accelerator accelerator,
+            Index3 extent)
+            where T : unmanaged => accelerator.AllocateExchangeBuffer<T>(extent, ExchangeBufferMode.PreferPagedLockedMemory);
+
+        /// <summary>
+        /// Allocates a new 3D exchange buffer that allocates the specified amount of
+        /// elements on the current accelerator. Furthermore, it keeps a buffer of the
+        /// same size in pinned CPU memory to enable asynchronous memory transfers
+        /// </summary>
+        /// <typeparam name="T">The element type.</typeparam>
+        /// <param name="accelerator">The associated accelerator to use.</param>
+        /// <param name="extent">The extent (number of elements to allocate).</param>
+        /// <param name="mode">The exchange buffer mode to use.</param>
+        /// <returns>The allocated 3D exchange buffer.</returns>
+        public static ExchangeBuffer3D<T> AllocateExchangeBuffer<T>(
+            this Accelerator accelerator,
+            Index3 extent,
+            ExchangeBufferMode mode)
+            where T : unmanaged
+        {
+            var gpuBuffer = accelerator.Allocate<T>(extent);
+            return new ExchangeBuffer3D<T>(gpuBuffer, mode);
+        }
```

Comment:
Although these methods seem fine to me I am wondering whether we could come up with a T4 template to generate these classes "automagically"?

Reply:
I don't have experience with T4 templates, I would have to learn about those.